concourse/steps/release.py

```diff
@@ -23,22 +29,317 @@
 )
 
 
-def rebase(
-    git_helper,
-    upstream_ref,
-):
-    upstream_commit_sha = git_helper.fetch_head(upstream_ref).hexsha
-    git_helper.rebase(commit_ish=upstream_commit_sha)
+class ReleaseContext(object):
+    def set_release_commit_sha(self, release_commit_sha):
+        self._release_commit_sha = release_commit_sha
+
+    def release_commit_sha(self):
+        return self._release_commit_sha
+
+    def set_release_notes(self, release_notes):
+        self._release_notes = release_notes
+
+    def release_notes(self):
+        return release_notes
+
+
+class TransactionalStep(object, metaclass=abc.ABCMeta):
+    ''' Provides a framework to implement some action in an atomic fashion.
+
+    Instances of this class are intended to perform exactly one change in their apply-method
+    and provide the corresponding undo-functionality in the revert-method
+    '''
+    def __init__(self, release_context: ReleaseContext):
+        self.release_context = release_context
+
+    @abc.abstractmethod
+    def validate(self):
+        pass
+
+    @abc.abstractmethod
+    def apply(self):
+        pass
+
+    @abc.abstractmethod
+    def revert(self):
+        pass
+
+    @abc.abstractmethod
+    def name(self):
+        pass
+
+
+class ReleaseCommitStep(TransactionalStep):
+    def __init__(
+        self,
+        git_helper: GitHelper,
+        repo_dir: str,
+        release_version: str,
+        repository_version_file_path: str,
+        repository_branch: str,
+        rebase_before_release: bool,
+        release_context: ReleaseContext,
+        release_commit_callback: str=None,
+    ):
+        super().__init__(release_context)
+        self.repository_version_file_path = existing_file(
+            pathlib.Path(repo_dir, repository_version_file_path)
+        )
+        self.git_helper = not_none(git_helper)
+        self.repo_dir = existing_dir(repo_dir)
+        self.repository_branch = repository_branch
+        self.release_version = not_empty(release_version)
+        self.rebase_before_release = rebase_before_release
+        self.release_commit_callback = release_commit_callback
+
+    def name(self):
+        return "Release Commit Step"
+
+    def validate(self):
+        info(f'validating {self.name()}')
+         # ensure the given callback script exists
+        if self.release_commit_callback:
+            self.release_commit_callback = os.path.join(
```

Comment:
no! this is a very very bad idea. either omit this in the validation, or mv the code to a method that returns the next version. then call it in `validate` w/o storing it. and call it again in `apply` (persisting it this time).

In case I missed other ocurrences: never ever modify the step object (or anything else) during validation. This is super-bad! Any step *must* successfully execute (at least potentially) w/o running validate before